Add GPS Lap Timer

SpencerGraffunder opened the pull request:
Resubmitting #10968. See discussion there.
Feature described here: https://youtu.be/TA5cWwFafY4

Test with:
make XYZ EXTRA_FLAGS="-DUSE_GPS -DUSE_GPS_LAP_TIMER"

ledvinap commented on src/main/osd/osd_elements.c:
```diff
@@ -1014,8 +1015,18 @@ static void osdElementGpsHomeDirection(osdElementParms_t *element)
 {
     if (STATE(GPS_FIX) && STATE(GPS_FIX_HOME)) {
         if (GPS_distanceToHome > 0) {
```
(technically, gpsLapTimerData.timerRunning will be ignored with GPS_distanceToHome == 0)

Also, does code below handle correctly case when distance to gate is zero?

ledvinap commented on src/main/osd/osd_elements.c:
```diff
@@ -1091,6 +1111,29 @@ static void osdElementEfficiency(osdElementParms_t *element)
+#ifdef USE_GPS_LAP_TIMER
+static void osdElementGpsLapTimeCurrent(osdElementParms_t *element)
+{
+    uint32_t lapTimeSeconds = gpsLapTimerData.currentLapTime / 1000;
```
(see above; prefix variant)

Also, isn't `-:--` possible here too ?

blckmn commented:
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> FAIL
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> FAIL
 - approver count at least three -> FAIL

SpencerGraffunder commented on src/main/target/common_pre.h:
```diff
@@ -373,7 +373,7 @@ extern uint8_t _dmaram_end__;
 #define USE_CANVAS
 #define USE_FRSKYOSD
 #define USE_GPS
-#define USE_GPS_NMEA
```
I agree. This was done mostly just let me test. We will have to discuss feature priority. I understand keeping NMEA for whacky receivers that can't do UBLOX. How do we decide which features go away when making room for new ones?

SpencerGraffunder commented:
Thanks for the through review, @blckmn! Just what I've been waiting for.
I'll have to take some time to work through your comments.

SpencerGraffunder commented on src/main/io/gps.c:
```diff
@@ -683,7 +685,17 @@ void gpsInitUblox(void)
                         break;
                     case 12:
-                        ubloxSetNavRate(0x64, 1, 1); // set rate to 10Hz (measurement period: 100ms, navigation rate: 1 cycle) (for 5Hz use 0xC8)
+                        switch (gpsConfig()->gps_update_rate) {
```
I think I want to just use an int to store the update rate.
The easiest way to do that would be to store it as a _period_ in milliseconds, because that can be passed straight to the existing `ubloxSetNavRate` function.
The other way, which would be easier for the user, would be to store it as a rate in Hz and have `ubloxSetNavRate` convert the Hz to a period. The problem here is that the rates may not be integers. The max rate I found to work on my units was 53ms, or 18.86Hz. So then we would set the max allowed value to 19Hz, but clamp the calculated period to 53ms.

ledvinap replied:
IMO using enum is a bit better here. It will prevent users using values that don't work. Without error message, this can be support nightmare.

SpencerGraffunder commented on src/main/osd/osd_elements.c:
```diff
@@ -1014,8 +1015,18 @@ static void osdElementGpsHomeDirection(osdElementParms_t *element)
 {
     if (STATE(GPS_FIX) && STATE(GPS_FIX_HOME)) {
         if (GPS_distanceToHome > 0) {
+#ifdef USE_GPS_LAP_TIMER
+            // Change the "home" point to the start/finish location if the lap timer is running
+            if (gpsLapTimerData.timerRunning) {
+                const int h = gpsLapTimerData.dirToPoint/100 - DECIDEGREES_TO_DEGREES(attitude.values.yaw);
+                element->buff[0] = osdGetDirectionSymbolFromHeading(h);
+            } else {
+#endif
             const int h = DECIDEGREES_TO_DEGREES(GPS_directionToHome - attitude.values.yaw);
             element->buff[0] = osdGetDirectionSymbolFromHeading(h);
+#ifdef USE_GPS_LAP_TIMER
+            }
+#endif
```
Definitely need to fix this section. Is there any reason not to do it like this? It uses a non-const int, but I assume that doesn't make it to the machine code anyways, right?

```suggestion
            int direction = GPS_directionToHome
#ifdef USE_GPS_LAP_TIMER
            // Override the "home" point to the start/finish location if the lap timer is running
            if (gpsLapTimerData.timerRunning) {
                direction = gpsLapTimerData.dirToPoint/10; // dirToPoint is centidegrees
            }
#endif
            element->buff[0] = osdGetDirectionSymbolFromHeading(DECIDEGREES_TO_DEGREES(direction - attitude.values.yaw));
```

ledvinap replied:
Yes, overriding direction is fine. 
My comment was mainly about two #ifdef mixed into C block, it is too easy to break with future changes.

ledvinap commented:
BTW: What about using home position as a gate? Gate positioning code would apply to home (which may be useful for other purposes), only problem I see is unintended flyaway when someone saves different site as home (I heard of $20k drone flying back to Africa on first test)

SpencerGraffunder commented:
> BTW: What about using home position as a gate? Gate positioning code would apply to home (which may be useful for other purposes)

Can you explain more? I don't see how it would be helpful.

ledvinap commented:
Current code uses concept of 'home' location. It is initialized on arming and used for HOME info on OSD and RTH code. But it is quite similar to your gate location. 
If user can specify Home location (using mechanisms from this PR), it can be used as Gate. And it will allow specifying RTH/OSD home that is different from arming location (possibly useful if you want clear place for RTH / OSD landing). Eventually, it can be extended to support geofencing etc.

SpencerGraffunder commented:
> Current code uses concept of 'home' location. It is initialized on arming and used for HOME info on OSD and RTH code. But it is quite similar to your gate location. If user can specify Home location (using mechanisms from this PR), it can be used as Gate. And it will allow specifying RTH/OSD home that is different from arming location (possibly useful if you want clear place for RTH / OSD landing). Eventually, it can be extended to support geofencing etc.

I see what you're talking about now. I think a non-volatile RTH position wouldn't be desirable because of flyaways. I could see it being cool to have a completely separate "POI" with a distance and pointer that you could have set at your vacation home or an airport, but I don't feel like betaflight is the right place for a secondary airfield feature.
If we don't like the idea of hijacking the home OSD elements, I could create a second set that's just for the lap timer.

ctzsnooze commented:
Very cool!

Thanks @ledvinap for your assistance.

Can I clarify that GPS Rescue can still be activated via switch or RC loss, while this function is active?  I can imagine this opening up the possibility of relatively long-distance races where a functional GPS Rescue could be required, even if the user wants to use GPS to time their laps.

Also, GPS Rescue now does good landings.  I can imagine that if the user chose the 'set home point once' rescue option, and arms once at the landing point, then they could go racing with this, do their laps, and at the end just hit GPS Rescue switch, and it would bring back and land at the correct spot.  Hence I think that keeping the GPS Rescue capability and its home point separate, and both functional, would be good.

How much extra flash space is required?

SpencerGraffunder commented:
> 

I haven't tested GPS Rescue while the timer is running, but I can't imagine any way it would interfere. I'm not planning to overwrite the home point at all other than hijacking the OSD element. Even that isn't necessary if we want to make two new elements for distance and direction to the gate. I'm not controlling flight or injecting positioning data to existing structs.

Not sure on flash space. After implementing some of the suggestions here, I was able to build for all targets within the flash limits without removing any other features.

Unfortunately I am too busy to work on this any more this month. The code still needs some changes done, so this isn't ready for testing yet.

ctzsnooze commented:
OK thanks for the heads up.  We will put it 'on ice' until you get back to us.

KarateBrot commented on src/main/cms/cms_menu_gps_lap_timer.c:
```diff
@@ -0,0 +1,131 @@
+#include "cms/cms_menu_gps_lap_timer.h"
+
+#include "config/feature.h"
+
```
```suggestion
```

KarateBrot commented on src/main/fc/gps_lap_timer.c:
```diff
@@ -0,0 +1,167 @@
+#include "platform.h"
+
+#ifdef USE_GPS_LAP_TIMER
+
```
```suggestion

#include "common/maths.h"

```

SpencerGraffunder replied:
It seems like platform.h is needed to pass the automated testing. Why do you suggest removing it?
Also, just curious why isn't the #ifdef isn't needed here?

SpencerGraffunder replied:
@KarateBrot

KarateBrot replied:
No, I just suggested to add `#include "common/maths.h"` so that you can get rid of the `MIN` and `MAX` macros down below (because both macros are defined in `maths.h`)

KarateBrot commented on src/main/fc/gps_lap_timer.c:
```diff
@@ -0,0 +1,167 @@
+    gpsLapTimerConfigMutable()->gateLon = gpsLapTimerData.gateLocation.lon;
+}
+
+#define MIN(a,b) (((a)<(b))?(a):(b))
+#define MAX(a,b) (((a)>(b))?(a):(b))
+
```
```suggestion
```

These macros are defined in `common/maths.h`, so we can remove them here.

KarateBrot commented:
>  Later on we need to test and confirm that GPS Rescue still works while this is active.

Why would GPS Rescue not work with this being active? This is the normal acro mode we are used to, just with an OSD element for lap times that is using the GPS data to get times. This is not some fancy flight mode that could potentially block other flight modes.

howels commented:
@SpencerGraffunder can you update this so it's ready for 4.5? Looks like some changes were requested but the feature freeze was missed. However it's a great looking feature so please update as I'm sure this will be popular.

KarateBrot commented:
> @SpencerGraffunder can you update this so it's ready for 4.5? Looks like some changes were requested but the feature freeze was missed. However it's a great looking feature so please update as I'm sure this will be popular.

Feature freeze wasn't missed. This PR was opened well before the feature freeze so no problem 👍

howels commented:
> > @SpencerGraffunder can you update this so it's ready for 4.5? Looks like some changes were requested but the feature freeze was missed. However it's a great looking feature so please update as I'm sure this will be popular.
> 
> Feature freeze wasn't missed. This PR was opened well before the feature freeze so no problem +1

thanks for the correction, was worried this wasn't making the cut. Even better news!

haslinghuis commented:
@SpencerGraffunder this need a rebase to resolve conflict

KarateBrot commented:
@haslinghuis This PR needs a bit of time and love because right now something is not working quite right. @SpencerGraffunder  says going through the gate doesn't get recognized. Probably a very small error somewhere that comes from converting the units from "cm" to "m".

sugaarK commented:
If we are releasing rc1 might have to bump this too 4.5 @KarateBrot @SpencerGraffunder

SpencerGraffunder commented:
I rebased. The configurable gps update rate was removed in the process. Could be added back in but I couldn't easily figure out where the gps config moved to. Logic didn't change so still not working. Anyone who cares about this feature is welcome to look into it. I might get some time this weekend. If it needs to be bumped to 4.5, I'm fine with that.

KarateBrot commented:
@SpencerGraffunder I can try to rebase and fix it.

KarateBrot commented:
@SpencerGraffunder Rebased. You can now fetch and pull and everything should be in order :)

mr-sour commented:
I gave this a rebase over here https://github.com/mr-sour/sourflight/tree/gps-lap-timer and a test. I couldn't set the home gate or get a "reading" so the rebases have broken something. I was able to kinda manually kick it with the cli and see `###ERROR IN get: CORRUPTED CONFIG: gps_lap_timer_gate_lon = -940783240###` I think the ### being the culprit of the issue.

mr-sour commented:
Quick update, got the home reading working. I didn't have reset timer bound to anything so thats why it wouldn't register. But as for the rest of it I couldn't trigger a lap nor do I have OSD elements so that bears some investigation.

KarateBrot commented:
@nerdCopter I think there was a problem (maybe a unit error) somewhere. As far as I remember, a lap doesn't register if you fly through the finish gate.

SpencerGraffunder commented:
I rebased to master, did mr-sour's lat-lng limit change, and did KarateBrot's changes. Haven't tested it yet. Ever since I moved to the mountains it takes like 10 minutes to get a GPS lock.

@mr-sour Were you running on an F411 or low-flash chip? As is, this feature is only enabled with TARGET_FLASH_SIZE > 512. I've been temporarily replacing one of the F411 defines with #define USE_GPS_LAP_TIMER whenever I test on a F411 FC.

mr-sour commented:
I was trying to test with a f405, so all good there. I think it was a matter of having a poor home gate. Maybe it drifted to much? Or the averaging when setting the home gate failed me and it averaged to the wrong spot. Perhaps we should set a minimum GPS sat count ( or use failsafe count)  for calculations going into the averaging? I think that's what was messing it up.

haslinghuis commented on src/main/target/common_pre.h:
```diff
@@ -370,6 +370,8 @@
 #define USE_GPS_NMEA
 #define USE_GPS_UBLOX
 #define USE_GPS_RESCUE
+#define USE_GPS_LAP_TIMER
+#define USE_CMS_GPS_LAP_TIMER_MENU
```
These should be optional or custom defines in build API.
```suggestion
#if defined (USE_GPS_LAP_TIMER)
#define USE_CMS_GPS_LAP_TIMER_MENU
#endif
```
With this no changes needed in common_pre.h anymore.

haslinghuis replied:
With BFC 10.9 / 10.10 user can add this using Custom Defines
But we also can ask @blckmn to add it to Other Options.

haslinghuis replied:
![image](https://user-images.githubusercontent.com/8344830/236288826-397f33cb-3757-4218-abcc-19b8197bcb4d.png)

SpencerGraffunder commented:
> I was trying to test with a f405, so all good there. I think it was a matter of having a poor home gate. Maybe it drifted to much? Or the averaging when setting the home gate failed me and it averaged to the wrong spot. Perhaps we should set a minimum GPS sat count ( or use failsafe count) for calculations going into the averaging? I think that's what was messing it up.

I assumed the F411 thing because of what you said about the OSD elements. Did you turn on the "unknown" elements in the configurator?

I don't care for minimum sat limits because sometimes I just can't get enough and sometimes it just doesn't matter for what I'm doing. I could get onboard with a text warning to retry when setting the gate if we notice the current reading is getting too far away from the average.

mr-sour commented:
Makes sense to me. Maybe a GPS_NOT_WARM warning then? I'll test the rebase tonight with a gps that's been settled in for 24 and hopefully I'll be able to run a lap.  I also have a analog quad I will try to test on to see if it's my digital osd mucking with the unknown messages.

SpencerGraffunder commented:
Finally got the chance to test it! It works!!!
Except the OSD elements don't show at all. I can only tell it's working by the end screen stats, which do show fine.
That's all I'll have time for today, so if anyone wants to look into the OSD code, that's what we need to fix. It should show 0's as long as there's a GPS connected I think so it can be tested on the bench.

SpencerGraffunder commented:
I'm working on making OSD elements for this feature. The OSD elements will have to go into a configurator pull request also. Can I get a definite answer on whether I can overwrite a few OSD elements? There are 4 that aren't used in osd_elements.h,  don't line up with an ASCII character, and I don't see them used in the code in any way (0x24, 0x77, 0x78, 0x79). The .mcm file has them as ![image](https://user-images.githubusercontent.com/17893168/236253816-1199fd63-2ca5-43cc-ac1c-07022b73c016.png), ![image](https://user-images.githubusercontent.com/17893168/236254025-c0f2e91a-eb86-4272-8458-17ec33858e28.png), ![image](https://user-images.githubusercontent.com/17893168/236254082-5931a5ea-ab7e-4fa3-a983-4213aeab29ce.png), and blank. The small arrows could be nice to keep around.
I can reuse the SYM_TOTAL_DISTANCE ![image](https://user-images.githubusercontent.com/17893168/236254828-65bcae85-842a-4123-af4a-4372f00108ed.png) as current lap time, so I need two more.
I would like to use 0x24 as a checkered flag (copied from inav) for best 3 and 0x79 as a stopwatch for previous lap.
![image](https://user-images.githubusercontent.com/17893168/236253367-9c723f19-065b-4170-93bb-5b0e2a2f503f.png)![image](https://user-images.githubusercontent.com/17893168/236254382-ce8205f1-34f4-462c-b438-cf27633b5c8b.png)

ledvinap commented on src/main/osd/osd_elements.c:
```diff
@@ -1040,8 +1041,14 @@ static void osdElementGpsHomeDirection(osdElementParms_t *element)
+#ifdef USE_GPS_LAP_TIMER
+            // Override the "home" point to the start/finish location if the lap timer is running
+            if (gpsLapTimerData.timerRunning) {
+                direction = gpsLapTimerData.dirToPoint/10; // dirToPoint is centidegrees
```
```suggestion
                direction = gpsLapTimerData.dirToPoint / 10; // dirToPoint is centidegrees
```

ledvinap commented on src/main/osd/osd_elements.c:
```diff
@@ -1056,8 +1063,16 @@ static void osdElementGpsHomeDirection(osdElementParms_t *element)
 
 static void osdElementGpsHomeDistance(osdElementParms_t *element)
 {
+    osdFormatDistanceString(element->buff, gpsLapTimerData.currentLapTime/1000, SYM_HOMEFLAG);
```
```suggestion
    osdFormatDistanceString(element->buff, gpsLapTimerData.currentLapTime / 1000, SYM_HOMEFLAG);
```

ledvinap commented on src/main/osd/osd.c:
```diff
@@ -545,6 +548,18 @@ void osdInit(displayPort_t *osdDisplayPortToUse, osdDisplayPortDevice_e displayP
+#ifdef USE_GPS_LAP_TIMER
+void printLapTime(char* buffer, uint32_t time) {
+    if (time != 0) {
+        const int timeSeconds = time / 1000;
```
(you can gain slightly better precision by rounding  - `time = (time + 5) / 10` )

ledvinap commented on src/main/io/gps.c:
```diff
@@ -543,10 +544,14 @@ static void ubloxSetMessageRate(uint8_t messageClass, uint8_t messageID, uint8_t
     ubxMessage_t tx_buffer;
-    tx_buffer.payload.cfg_rate.measRate = measRate;
+
+    uint8_t measRateMilliseconds = 1000/measRateHz;
```
```suggestion
    uint8_t measRateMilliseconds = 1000 / measRateHz;
```

Also, `measRateHz < 4 will overflow

ledvinap reviewed (approved):
Not tested, but code looks solid

SpencerGraffunder commented:
I took a few hours off work today to try to make the OSD elements display but was unsuccessful. I even tried making a copy of one of the other existing elements and that wouldn't display either. I'm at a loss. I'm going to need some help here.

nerdCopter commented:
`SYM_THR` is one of the most basic ones you can inspect.
grep the code and find 
```
$ grep -Iirn SYM_THR src/ 
src/main/osd/osd_elements.c:1524:    tfp_sprintf(element->buff, "%c%3d", SYM_THR, calculateThrottlePercent());
src/main/drivers/osd_symbols.h:48:#define SYM_THR                     0x04
```
then grep the code for the display 
```
$ grep -Iirn osdElementThrottlePosition src/
src/main/osd/osd_elements.c:1522:static void osdElementThrottlePosition(osdElementParms_t *element)
src/main/osd/osd_elements.c:1756:    [OSD_THROTTLE_POS]            = osdElementThrottlePosition,
```
i think these lines/code/locations will lead to success.   `%c` is the character spot for SYM_THR and in this case, %3d is 3 digits for the value.

KarateBrot commented on src/main/fc/gps_lap_timer.c:
```diff
@@ -0,0 +1,158 @@
+#include <stdint.h>
+#include <math.h>
+
+#include "common/maths.h"
+
+#include "platform.h"
```
```suggestion
#include "platform.h"

#include "common/maths.h"
```

KarateBrot replied:
Just switch them around :)

SteveCEvans commented on src/main/osd/osd.h:
```diff
@@ -186,6 +186,9 @@ typedef enum {
     OSD_SYS_WARNINGS,
     OSD_SYS_VTX_TEMP,
     OSD_SYS_FAN_SPEED,
+    OSD_GPS_LAP_TIME_CURRENT,
```
Any value greater than or equal to OSD_SYS_GOGGLE_VOLTAGE is treated as a system element and won't be rendered. I suggest the following fix. 

```
diff --git a/src/main/osd/osd_elements.c b/src/main/osd/osd_elements.c
index ecc72f9..ea82f15 100644
--- a/src/main/osd/osd_elements.c
+++ b/src/main/osd/osd_elements.c
@@ -2016,7 +2016,7 @@ static void osdDrawSingleElement(displayPort_t *osdDisplayPort, uint8_t item)
     element.attr = DISPLAYPORT_SEVERITY_NORMAL;
 
     // Call the element drawing function
-    if ((item >= OSD_SYS_GOGGLE_VOLTAGE) && (item < OSD_ITEM_COUNT)) {
+    if ((item >= OSD_SYS_GOGGLE_VOLTAGE) && (item < OSD_GPS_LAP_TIME_CURRENT)) {
         displaySys(osdDisplayPort, elemPosX, elemPosY, (displayPortSystemElement_e)(item - OSD_SYS_GOGGLE_VOLTAGE + DISPLAYPORT_SYS_GOGGLE_VOLTAGE));
     } else {
         osdElementDrawFunction[item](&element);
```

SteveCEvans replied:
With the above change...
![image](https://github.com/betaflight/betaflight/assets/11480839/b7b32671-ae7b-4067-a606-675e7b00cf27)

KarateBrot commented on src/main/cms/cms_menu_gps_lap_timer.c:
```diff
@@ -0,0 +1,128 @@
+    {"--- GPS LAP TIMER ---", OME_Label,   NULL,          NULL},
+
+    {"SET POSITION",          OME_Funcall, cmsMenuChange, &cmsSetGateMenu},
+    {"GATE RADIUS(m)",       OME_UINT8,   NULL,          &(OSD_UINT8_t){&gpsLapTimerConfig_gateToleranceM, 1, 100, 1}},
```
```suggestion
    {"GATE RADIUS(m)",        OME_UINT8,   NULL,          &(OSD_UINT8_t){&gpsLapTimerConfig_gateToleranceM, 1, 100, 1}},
```

KarateBrot reviewed (changes requested):
I think we're getting very close now that the OSD elements work 🎉

SpencerGraffunder commented:
It's working! Video shows lap times while flying and in end screen stats. Will clean up code  tomorrow. Screenshot from Rotorhazard to compare. I think it's mostly accurate within 0.15 seconds.

https://youtu.be/tM6D7Ybi8t4
![image](https://github.com/betaflight/betaflight/assets/17893168/9e572310-05ed-44d1-b40d-a3e7ea9b1971)

SteveCEvans commented on src/main/osd/osd_elements.c:
```diff
@@ -1962,7 +2017,7 @@ static void osdDrawSingleElement(displayPort_t *osdDisplayPort, uint8_t item)
 
     // Call the element drawing function
-    if ((item >= OSD_SYS_GOGGLE_VOLTAGE) && (item < OSD_ITEM_COUNT)) {
+    if ((item >= OSD_SYS_GOGGLE_VOLTAGE) && (item < OSD_GPS_LAP_TIME_CURRENT)) {
```
Sorry to be picky, having suggested this change, but it would be cleaner expressed as `<= OSD_SYS_FAN_SPEED`.

KarateBrot commented:
Nice! I've noticed that the GPS lap timer laptime is always too high (rather than too high or too low equally). So maybe there is a little integer rounding problem and we can do a little better for the timing precision. Maybe we can get the precision down to +/-0.07s. I'll see if I can notice something.

KarateBrot commented on src/main/osd/osd.c:
```diff
@@ -545,6 +548,18 @@ void osdInit(displayPort_t *osdDisplayPortToUse, osdDisplayPortDevice_e displayP
 }
 
+#ifdef USE_GPS_LAP_TIMER
+void printLapTime(char* buffer, uint32_t time) {
+    if (time != 0) {
+        const int timeSeconds = time / 1000;
+        const int timeDecimals = ((time % 1000) + 5) / 10;
```
I think we should round to the nearest centisecond also for `timeSeconds` so that both variables are consistently rounded.
```suggestion
void printLapTime(char* buffer, uint32_t timeMs) {
    if (timeMs != 0) {
        timeMs += 5;  // round to nearest centisecond (+/- 5ms)
        const int timeSeconds = timeMs / 1000;
        const int timeDecimals = (timeMs % 1000) / 10;
```

nerdCopter commented:
is it possible to put symbols on stats screen as well? (or maybe devs would disfavor? :man_shrugging: )
mockup: 
![image](https://github.com/betaflight/betaflight/assets/56646290/09b59a74-351f-4d66-aa3e-1049a37f1502)

KarateBrot commented:
@nerdCopter the stats screen doesn't have OSD symbols. I guess it would be pretty distracting if all stats now have a symbol in front (which also decreases the number of chars you can display per line).

SpencerGraffunder commented:
> Nice! I've noticed that the GPS lap timer laptime is always too high (rather than too high or too low equally). So maybe there is a little integer rounding problem and we can do a little better for the timing precision. Maybe we can get the precision down to +/-0.07s. I'll see if I can notice something

I just flew another 10 laps with both systems and I don't come to the same conclusion. This example shows the opposite. The GPS lap timer is using the same timestamp for the end of one lap and the start of the next so the only way I know of that it could be wrong is if the actual displaying of the number is rounding wrong or the gps point is "drifting" (somewhat expected). Feel free to look at the displayed number division and modulus code but I think we have that right.

Also a good number to have around: the standard deviation between the timing systems is 0.135 seconds.

<img width="522" alt="image" src="https://github.com/betaflight/betaflight/assets/17893168/44ae364a-5801-430e-8462-6a816ec7c93f">

KarateBrot commented:
Well, maybe it was just coincidence that the values were too high (sample size was only 5 laps in the video).

KarateBrot commented on src/main/osd/osd.c:
```diff
@@ -545,6 +548,19 @@ void osdInit(displayPort_t *osdDisplayPortToUse, osdDisplayPortDevice_e displayP
 }
 
+#ifdef USE_GPS_LAP_TIMER
+void printLapTime(char* buffer, uint32_t time) {
+    if (time != 0) {
+        timeRound = time + 5;
+        const int timeSeconds = timeRound / 1000;
+        const int timeDecimals = (timeRound % 1000) / 10;
```
`time` with the postfix "Ms" gives a hint about what time format goes into the function. Also, when calling the function, a copy of the time input is made, so you can directly manipulate `timeMs`. Therefore you don't need the variable `timeRound`.
```suggestion
void printLapTime(char *buffer, uint32_t timeMs) {
    if (timeMs != 0) {
        timeMs += 5;
        const int timeSeconds = timeMs / 1000;
        const int timeDecimals = (timeMs % 1000) / 10;
```

KarateBrot commented on src/main/osd/osd.c:
```diff
@@ -545,6 +548,19 @@ void osdInit(displayPort_t *osdDisplayPortToUse, osdDisplayPortDevice_e displayP
 }
 
+#ifdef USE_GPS_LAP_TIMER
+void printLapTime(char *buffer, uint32_t timeMs) {
```
Mark `timeMs` as `const` to show to others that you want it to be immutable inside of the function body. Otherwise you could just use `timeMs` to increment the time by 5, as it's already a copy of the original input.
```suggestion
void printLapTime(char *buffer, const uint32_t timeMs) {
```

KarateBrot replied:
Sorry, I didn't want to annoy you with this detail. Maybe I did not fully grasp your intentions. I don't want this to seem like you absolutely need to do what I suggest. If you have strong feelings against this, just change it back to the original way you wanted it to be. I'll just approve, as the PR is super solid (always has been).

ledvinap commented on src/main/cli/cli.c:
```diff
@@ -4400,6 +4437,14 @@ STATIC_UNIT_TESTED void cliSet(const char *cmdName, char *cmdline)
+                    int32_t value = strtol(eqptr, NULL, 10);
+
+                    // INT32s are limited to being symmetric, so we test both bounds with the same magnitude
+                    if (value <= val->config.d32Max && value >= -val->config.d32Max) {
```
```suggestion
                    const uint32_t absValue = ABS(value);
                    if (absValue <= val->config.u32Max) {
```
twos complement signed value are nasty. Minimum negative value is -(2^31), but maximum is 2^31-1. So value >= -val->config.d32Max would refuse -(2^31) (it is impossible to set limit correctly).

SpencerGraffunder replied:
This is comparing to the max allowed value for this particular CLI item as set in settings.c:1054 (900000000 and 1800000000 in this case). It has nothing to do with INT32_MAX or 2^31. So this conditional is essentially:
`if (value <= 1800000000  && value >= -1800000000 ) {`

ledvinap commented on src/main/osd/osd.c:
```diff
@@ -545,6 +548,19 @@ void osdInit(displayPort_t *osdDisplayPortToUse, osdDisplayPortDevice_e displayP
+#ifdef USE_GPS_LAP_TIMER
+void printLapTime(char *buffer, const uint32_t timeMs) {
+    if (timeMs != 0) {
+        const uint32_t timeRoundMs = timeMs + 5;
```
```suggestion
        const uint32_t timeRoundMs = timeMs + 5;  // round value in division by 10
```

ledvinap reviewed (commented):
Just minor comments, code is nice

SpencerGraffunder commented on src/main/osd/osd_elements.c:
```diff
@@ -1962,7 +2017,7 @@ static void osdDrawSingleElement(displayPort_t *osdDisplayPort, uint8_t item)
 
     // Call the element drawing function
-    if ((item >= OSD_SYS_GOGGLE_VOLTAGE) && (item < OSD_ITEM_COUNT)) {
+    if ((item >= OSD_SYS_GOGGLE_VOLTAGE) && (item <= OSD_SYS_FAN_SPEED)) {
```
I agree but this is what Steve suggested. I see two alternatives:

1. Add a macro like this and use it in the if()
`#define IS_SYS_OSD_ELEMENT(item) (item >= OSD_SYS_GOGGLE_VOLTAGE) && (item <= OSD_SYS_FAN_SPEED)`

2. Somehow add sys elements to the top of the list and we only have to check one bound again. Would need configurator change and mess up all OSD settings when upgrading though(?)

haslinghuis replied:
We still need to resolve @SteveCEvans suggested change:

```
// Call the element drawing function
-    if ((item >= OSD_SYS_GOGGLE_VOLTAGE) && (item < OSD_ITEM_COUNT)) {
+    if ((item >= OSD_SYS_GOGGLE_VOLTAGE) && (item < OSD_GPS_LAP_TIME_CURRENT)) {
         displaySys(osdDisplayPort, elemPosX, elemPosY, (displayPortSystemElement_e)(item - OSD_SYS_GOGGLE_VOLTAGE + DISPLAYPORT_SYS_GOGGLE_VOLTAGE));
     } else {
         osdElementDrawFunction[item](&element);

```

SpencerGraffunder replied:
Yes. Does anyone have a solid suggestion? Should I add my macro?

KarateBrot replied:
That's my 2 cents https://github.com/betaflight/betaflight/pull/11856#discussion_r1199535967

KarateBrot replied:
@SpencerGraffunder I guess you can add your own macro somewhere where it is apparent, for example right below `osdStats_e`. I have no better idea either.

SteveCEvans replied:
I'd be happy with a macro as it's cleaner.

SpencerGraffunder commented on src/main/cli/cli.c:
```diff
@@ -873,6 +896,16 @@ static void cliSetVar(const clivalue_t *var, const uint32_t value)
             *(uint32_t *)ptr = workValue;
             break;
+
+        case VAR_INT32:
```
Are you suggesting to store the signed lat/lng values in a uint32? I can see that working with like an offset of half the max value, but then setting/getting through the CLI would be super whacky.

If I'm misunderstanding, then I'll need some more explanation. I don't have a super deep understanding of this section.

haslinghuis commented:
Please rebase

KarateBrot commented on src/main/osd/osd_elements.c:
```diff
@@ -1962,7 +2017,7 @@ static void osdDrawSingleElement(displayPort_t *osdDisplayPort, uint8_t item)
 
     // Call the element drawing function
-    if ((item >= OSD_SYS_GOGGLE_VOLTAGE) && (item < OSD_ITEM_COUNT)) {
+    if ((item >= OSD_SYS_GOGGLE_VOLTAGE) && (item <= OSD_SYS_FAN_SPEED)) {
         displaySys(osdDisplayPort, elemPosX, elemPosY, (displayPortSystemElement_e)(item - OSD_SYS_GOGGLE_VOLTAGE + DISPLAYPORT_SYS_GOGGLE_VOLTAGE));
```
@ledvinap @SpencerGraffunder Right now the problem is that the `OSD_SYS_x` elements are blocking other elements from getting added to the end of the enumeration, I reckon.

![grafik](https://github.com/betaflight/betaflight/assets/19867640/8ce1cd02-5823-40c3-995d-c64d047a4ae2)

So I'd suggest changing the `osd_items_e` enums like this to always put `OSD_SYS_x` elements last:

```c
typedef enum {
    ...

    OSD_GPS_LAP_TIME_CURRENT,
    OSD_GPS_LAP_TIME_PREVIOUS,
    OSD_GPS_LAP_TIME_BEST3,
    OSD_ITEM_COUNT_VISIBLE,  // MUST BE LAST (visible item)

    // OSD_SYS_XXX items are rendered in the goggles firmware
    OSD_SYS_GOGGLE_VOLTAGE = OSD_ITEM_COUNT_VISIBLE,
    OSD_SYS_VTX_VOLTAGE,
    OSD_SYS_BITRATE,
    OSD_SYS_DELAY,
    OSD_SYS_DISTANCE,
    OSD_SYS_LQ,
    OSD_SYS_GOGGLE_DVR,
    OSD_SYS_VTX_DVR,
    OSD_SYS_WARNINGS,
    OSD_SYS_VTX_TEMP,
    OSD_SYS_FAN_SPEED,
    OSD_ITEM_COUNT  // MUST BE LAST (sys item)
} osd_items_e;
```

Then we can solve this a bit more cleanly and we don't need to touch it anymore when adding new osd items:

```suggestion
    if ((item >= OSD_ITEM_COUNT_VISIBLE) && (item < OSD_ITEM_COUNT)) {
        displaySys(osdDisplayPort, elemPosX, elemPosY, (displayPortSystemElement_e)(item - OSD_ITEM_COUNT_VISIBLE + DISPLAYPORT_SYS_GOGGLE_VOLTAGE));
```

SpencerGraffunder replied:
We wouldn’t have to change those lines at all after this if new elements keep getting added to the end like they’re supposed to. I assume shifting the sys elements down every time an element is added would also require more configurator integration. Correct me if I’m wrong though. Maybe the solution would be just a comment that explains what’s going on there. There’s nothing majorly wrong with it, it’s just confusing if you don’t already know because it’s not self explanatory.

KarateBrot replied:
Yeah, I have not thought about the configurator part. You are probably right. Forget about this suggestion.

haslinghuis commented:
@SteveCEvans are the requested changes resolved?

haslinghuis commented on src/main/cms/cms_menu_gps_lap_timer.c:
```diff
@@ -0,0 +1,128 @@
+/*
+ * This file is part of Cleanflight and Betaflight.
+ *
+ * Cleanflight and Betaflight are free software. You can redistribute
+ * this software and/or modify this software under the terms of the
+ * GNU General Public License as published by the Free Software
+ * Foundation, either version 3 of the License, or (at your option)
+ * any later version.
+ *
+ * Cleanflight and Betaflight are distributed in the hope that they
+ * will be useful, but WITHOUT ANY WARRANTY; without even the implied
+ * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this software.
+ *
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
```
Do we need updated license?

haslinghuis replied:
See [src/main/fc/gps_lap_timer.h](https://github.com/betaflight/betaflight/pull/11856/files#diff-acfa8e57f55d7effc8d629851e13d0c8ca7108aba9ef09df4d1bcd3988e7095e)

KarateBrot replied:
Good call, forgot to mention it

KarateBrot commented on src/main/fc/gps_lap_timer.c:
```diff
@@ -0,0 +1,150 @@
+        } else {
+            // Just left the circle, so record the lap time
+            if (wasInCircle) {
+                // Not the first time through the gate
```
If you have a compass you could dial in the heading of the gate directly through the CMS menu or CLI

ledvinap replied:
@SteveCEvans IIRC gate time is taken at closest point to gate (minimum distance is tracked, with timestamp). So actual gate size should not matter.

SpencerGraffunder replied:
An exit direction could help by letting us make the circle bigger and not be caught by other parts of the track. Since it's using the closest point to the gate waypoint, I think that's the most accurate we can be without correcting the gate waypoint. I'm sure there is a lot more we could do with this in the future. Split times and auto-correcting the start/finish point are two things I've thought of.

SpencerGraffunder commented on src/main/cms/cms_menu_gps_lap_timer.c:
```diff
@@ -0,0 +1,128 @@
+/*
+ * This file is part of Cleanflight and Betaflight.
+ *
+ * Cleanflight and Betaflight are free software. You can redistribute
+ * this software and/or modify this software under the terms of the
+ * GNU General Public License as published by the Free Software
+ * Foundation, either version 3 of the License, or (at your option)
+ * any later version.
+ *
+ * Cleanflight and Betaflight are distributed in the hope that they
+ * will be useful, but WITHOUT ANY WARRANTY; without even the implied
+ * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this software.
+ *
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
```
Which ones are the old and new license then? This is the same as in gps_lap_timer.c and most of the other files.

haslinghuis commented on src/main/osd/osd_elements.c:
```diff
@@ -1129,6 +1143,34 @@ static void osdElementEfficiency(osdElementParms_t *element)
 #endif // USE_GPS
 
+#ifdef USE_GPS_LAP_TIMER
+static void osdFormatLapTime(osdElementParms_t *element, uint32_t timeMS, uint8_t symbol)
+{
+    uint32_t seconds = timeMS / 1000;
+    uint32_t decimals = (timeMS % 1000) / 10;
```
Skipped

haslinghuis commented:
Like to have a test result as `Testing Required` label prevents this from merging.

- Do we need changes in configurator?
- Do we need documentation how to use the feature?

![image](https://github.com/betaflight/betaflight/assets/8344830/c988bb1c-74ca-43d3-bfc9-2deb266eeceb)

Easy4Racing commented:
Hi , I want to run a test for this new feature but failed. 
1. I flashed the new Hex file cloud built by Configurator(version 10.10.0-debug-e6f3b5b). 
2. Selected sensor input 'GPS' in Ports tab, enabled GPS & OSD in Configuration tab, enabled 'Unknown 1.2.3' in OSD tab. Nothing of GPS_lap_timer displayed but the craft_name and on time is ok. 

What should I do next ?
My board is OMNIBUSF4

SpencerGraffunder commented:
> What should I do next ? My board is OMNIBUSF4

Did you add USE_GPS_LAP_TIMER in custom defines? 

![image](https://github.com/betaflight/betaflight/assets/17893168/b2b0bff5-0a2c-4204-84b6-70f3e6486c83)


Your GPS also needs a lock before the elements will show.

Easy4Racing commented:
> > What should I do next ? My board is OMNIBUSF4
> 
> Did you add USE_GPS_LAP_TIMER in custom defines?
> 
> ![image](https://user-images.githubusercontent.com/17893168/241472316-b2b0bff5-0a2c-4204-84b6-70f3e6486c83.png)
> 
> Your GPS also needs a lock before the elements will show.

Thx for reply.
I set the custom defines but have not connect the GPS module. So, this is the reason. 
I thought here may be an initialization value, even if I don't pick up GPS, or the GPS module work improperly .

![W1V2Y3P1NKSTQ5E4`Z{IVJP](https://github.com/betaflight/betaflight/assets/51906813/79d938ea-603c-4b48-8fc2-04582eb26745)

![firmware](https://github.com/betaflight/betaflight/assets/51906813/6c4f103c-7334-4fd7-b3b0-ea56bfd99e0a)

haslinghuis commented:
Don't use `USE` in Custom Defines: `GPS_LAP_TIMER`
